### PR TITLE
Support NextN = 2/4 in DSV32

### PR DESCRIPTION
**Pull request by b8zhong**
## Motivation

Ref: https://github.com/deepseek-ai/DeepGEMM/pull/316: it removes the multicast in this kernel, allowing any next_n ≥ 1 on SM100 only.

## Modifications

Support it. Note: on SM90 it currently support `nextn=2`. But I don't have machine, so didn't test it. Just add a TODO

## Accuracy Tests

GPQA for V3.2 NVFP4:
```
python3 -m sglang.test.run_eval --port 30000 --eval-name gpqa --num-examples 198 --max-tokens 128000 --repeat 8 --top-p 0.95 --temperature 1.0 --thinking-mode deepseek-v3
```

Before:
```
Repeat: 8, mean: 0.821███████████████████████████████████████████████████████████████████████████████████████████████                 | 172/198 [48:35<07:33, 17.46s/it]
Scores: ['0.828', '0.825', '0.825', '0.828', '0.828', '0.812', '0.811', '0.811']██████████████████████████████████████████████▊       | 187/198 [49:42<01:32,  8.43s/it]
Mean latency: 2877.289 s███████████████████████████████████████████████████████████████████████████████████▉                          | 158/198 [48:24<17:23, 26.08s/it]
```

After (NextN = 2)
```
Repeat: 8, mean: 0.837███████████████████████████████████████████████████████████████████████████████████████████████████████▌        | 185/198 [52:34<04:21, 20.11s/it]
Scores: ['0.864', '0.828', '0.838', '0.833', '0.833', '0.848', '0.838', '0.813']█████████████████████████████████████                 | 172/198 [51:07<04:56, 11.41s/it]
Mean latency: 3009.260 s
```

After (NextN = 4)
```
Repeat: 8, mean: 0.838████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎          | 187/198 [49:47<03:34, 19.46s/it]
Scores: ['0.859', '0.828', '0.848', '0.864', '0.848', '0.838', '0.798', '0.818']
Mean latency: 2823.050 s█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋                       | 174/198 [48:56<07:57, 19.88s/it]
====================██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏      | 191/198 [49:04<00:57,  8.25s/it]
Output throughput: 493.361 token/s
```

Acceptance leeength (has no change):
Before:
```
+-------------+--------+------------+-----------------+
| Latency (s) | Tokens | Acc Length | Speed (token/s) |
+-------------+--------+------------+-----------------+
|    4.090    |  1024  |   2.860    |     250.36      |
+-------------+--------+------------+-----------------+
```

After:
```
+-------------+--------+------------+-----------------+
| Latency (s) | Tokens | Acc Length | Speed (token/s) |
+-------------+--------+------------+-----------------+
|    4.055    |  1024  |   2.860    |     252.55      |
+-------------+--------+------------+-----------------+
```

## Speed Tests and Profiling

```
python3 -m sglang.launch_server \
  --model-path nvidia/DeepSeek-V3.2-NVFP4 \
  --tp 8 \
  --ep 8 \
  --quantization modelopt_fp4 \
  --speculative-algorithm EAGLE \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --tool-call-parser deepseekv32 \
  --reasoning-parser deepseek-v3
```
```
SGLANG_TORCH_PROFILER_DIR="./" \
SGLANG_PROFILE_RECORD_SHAPES=true \
SGLANG_PROFILE_WITH_STACK=true \
python3 -m sglang.bench_one_batch_server \
    --model baseten-admin/glm-4.7-fp8-attn-fp4-mlp \
    --base-url http://localhost:30000/ \
    --batch-size 32 \
    --input-len 32768 \
    --output-len 256 \
    --profile \
    --profile-steps 10 \
    --show-report \
    --profile-by-stage
```

E2E example: 2940.74 -> 3157.50 decode TPS for context = 32K, BS = 32 under MTP (around 7%)

Before:
Screenshot 2026-05-09 at 11 30 48 PM

After:
Screenshot 2026-05-09 at 11 25 16 PM

Note that the kernel name has changed to `deep_gemm::sm100_fp8_paged_mqa_logits<4u, 64u, 128u, 64u, true, false, 3u, 4u, 256u, 128u, 256u, float, 2u>`, notice the change from `1u` to `4u`, which can confirm it's being used correctly.

Unit benchmark with CUPTI:
### next_n = 2

| B \ ctx |  2K  |  4K  |  8K  |  16K |  32K |  64K |
|--------:|:----:|:----:|:----:|:----:|:----:|:----:|
| 1       | 1.01 | 0.96 | 0.98 | 0.96 | 1.02 | 1.05 |
| 2       | 1.01 | 0.97 | 0.95 | 1.03 | 1.03 | 1.05 |
| 4       | 0.96 | 0.98 | 0.98 | 1.03 | 1.05 | 1.10 |
| 8       | 0.94 | 1.01 | 1.00 | 0.96 | 1.05 | 1.22 |
| 16      | 1.03 | 1.04 | 1.06 | 1.08 | **1.26** | **1.54** |
| 32      | 1.01 | 1.02 | 1.09 | **1.26** | **1.45** | **1.59** |
| 64      | 0.97 | 1.03 | **1.16** | **1.20** | **1.27** | **1.70** |

### next_n = 3

| B \ ctx |  2K  |  4K  |  8K  |  16K |  32K |  64K |
|--------:|:----:|:----:|:----:|:----:|:----:|:----:|
| 1       | 0.99 | 0.96 | 0.97 | 1.04 | 0.99 | 0.92 |
| 2       | 0.98 | 0.98 | 1.03 | 0.95 | 0.97 | 0.97 |
| 4       | 0.92 | 0.96 | 0.98 | 0.97 | 0.99 | 1.02 |
| 8       | 0.96 | 0.99 | 1.01 | 1.05 | 0.97 | 1.06 |
| 16      | 1.00 | 0.98 | 0.93 | 0.92 | 0.98 | 0.86 |
| 32      | 0.99 | 1.00 | 0.97 | 1.06 | 1.09 | 1.09 |
| 64      | 1.03 | 1.06 | **1.13** | **1.31** | **1.57** | **1.45** |

### next_n = 4

| B \ ctx |  2K  |  4K  |  8K  |  16K |  32K |  64K |
|--------:|:----:|:----:|:----:|:----:|:----:|:----:|
| 1       | 0.97 | 0.90 | 0.94 | 1.01 | 1.07 | 1.05 |
| 2       | 0.93 | 0.97 | 1.05 | 1.05 | 1.05 | 1.06 |
| 4       | 0.92 | 0.99 | 1.01 | 1.05 | 1.06 | 1.12 |
| 8       | 0.99 | 1.04 | 1.04 | **1.12** | **1.19** | **1.31** |
| 16      | 1.02 | 1.04 | **1.09** | **1.13** | **1.25** | **1.39** |
| 32      | 1.07 | 1.05 | **1.11** | **1.21** | **1.32** | **1.74** |
| 64      | 1.11 | **1.18** | **1.29** | **1.56** | **1.96** | **1.85** |

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #26482763282](https://github.com/sgl-project/sglang/actions/runs/26482763282)
Latest PR Test (Extra): :x: [Run #26853200137](https://github.com/sgl-project/sglang/actions/runs/26853200137)

### CODE REVIEW

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
/rerun-test test/registered/models_e2e/test_dsa_dsv32_tp_mtp.py test/registered/models_e2e/test_dsa_dsv32_dp_mtp.py

**Comment by Fridge003** [comment]
Wait for running of extra test

**Comment by ch-wan** [comment]
I recently added some attention unittests, and the main branch failed to pass one DSA test after this PR was merged. I'm checking if reverting this PR recovers the test.
